src/phylo_flags.hpp

```diff
@@ -0,0 +1,240 @@
+  PhyloFlagName substitution_model_;
+  PhyloFlagName substitution_model_rates_;
+  PhyloFlagName substitution_model_frequencies_;
+  PhyloFlagName site_model_parameters_;
+  PhyloFlagName clock_model_rates_;
+  PhyloFlagName entire_substitution_model_;
+  PhyloFlagName entire_site_model_;
+  PhyloFlagName entire_clock_model_;
+  // All Gradient map keys.
+  std::vector<PhyloFlagName> all_keys_;
+};
+
+// ** All User Options
+class PhyloFlags {
+ public:
+  PhyloFlags(bool is_run_defaults = true);
+  PhyloFlags(const StringVector &key_vec, bool is_run_defaults = true);
+  PhyloFlags(const StringDoubleVector &key_vec, bool is_run_defaults = true);
+
+  // Find flag by name.
+  std::optional<PhyloFlagName> FindFlagByName(const std::set<PhyloFlagName> &options,
+                                              const std::string &flag_name,
+                                              std::optional<PhyloFlagName> flag) const;
+  // Find flag by id.
+  std::optional<PhyloFlagName> FindFlagById(const std::set<PhyloFlagName> &options,
+                                            const size_t &flag_id,
+                                            std::optional<PhyloFlagName> flag) const;
+
+  // Set flag option.
+  void SetFlag(const std::string &flag_name, const bool set = true,
+               const double value = 1.0f);
+  void SetFlag(const PhyloFlagName &flag, const bool set = true,
+               const double value = 1.0f);
+  // Set flag option with specified value.
+  void SetFlagWithValue(const std::string &flag_name, const double value);
+  void SetFlagWithValue(const PhyloFlagName &flag, const double value);
+
+  // Special case for the is_run_defaults key
+  void SetRunDefaultsFlag(bool is_set);
+  bool IsRunDefaultsFlagSet();
+
+  // Check if flag option has been added.
+  bool IsFlagSet(const std::string &flag_name) const;
+  bool IsFlagSet(const PhyloFlagName &flag) const;
+  bool IsFlagNotSet(const std::string &flag_name) const;
+  bool IsFlagNotSet(const PhyloFlagName &flag) const;
+
+  // Special flag for indicating all options to be considered true.
+  bool IsRunDefaultsSet() const;
+
+  // Returns the value of the flag.
+  std::optional<double> GetFlagValue(const std::string &flag_name) const;
+  std::optional<double> GetFlagValue(const PhyloFlagName &flag) const;
+  // Returns the flag's value if set, otherwise returns default value.
+  double GetFlagValueIfSet(const std::string &flag_name, double default_value) const;
+  double GetFlagValueIfSet(const PhyloFlagName &flag, double default_value) const;
+  static double GetFlagValueIfSet(const std::optional<PhyloFlags> phylo_flags,
+                                  const PhyloFlagName &flag, double default_value);
+
+  // Convert flag to boolean vector.
+  BoolVector ToBoolFlags(const StringVector &vec_of_keys) const;
+  // Interprets flags as a string.
+  std::string ToString() const;
+
+  // Checks if a flag has been set, or whether to use defaults.
```

Comment:
Can you say a little more here about the difference between "Checked" and "Set"?